mipengine/algorithms/histogram.py

```diff
@@ -0,0 +1,391 @@
+    # print(metadata)
+    vars = [var for var in xvars + yvars if var != "dataset"]
+
+    numerical_vars = [var for var in vars if not metadata[var]["is_categorical"]]
+    nominal_vars = [var for var in vars if metadata[var]["is_categorical"]]
+
+    enumerations_dict = {var: metadata[var]["enumerations"] for var in nominal_vars}
+    print(enumerations_dict)
+
+    yvar = yvars[0]
+    if yvar in xvars:
+        xvars.remove(yvar)
+    print(yvar)
+    print(xvars)
+    if yvar in nominal_vars:
+        locals_result = local_run(
+            func=compute_local_histogram_categorical_secure,
+            positional_args=[data, enumerations_dict, yvar, xvars],
+            share_to_global=[True],
+        )
+
+        categorical_histogram = get_transfer_data(
+            global_run(
+                func=merge_grouped_histogram_categorical_secure2,
+                positional_args=[locals_result, xvars],
+                share_to_locals=[False],
+            )
+        )
+        numerical_histogram = {}
+
+    else:
+        # find_min_max_local()
+        local_result4 = local_run(
+            func=find_min_max_local,
+            positional_args=[data, yvar],
+            share_to_global=[True],
+        )
+        # find_min_max_global()
+        min_max_result = get_transfer_data(
+            global_run(
+                func=find_min_max_global,
+                positional_args=[local_result4],
+                share_to_locals=[False],
+            )
+        )
+        min_value = min_max_result["min_value"]
+        max_value = min_max_result["max_value"]
+        # compute_local_histogram_numerical(data,bins,yvar,xvars,min_value,max_value)
+        local_result5, bins_transfer = local_run(
+            func=compute_local_histogram_numerical_secure2,
+            positional_args=[
+                data,
+                enumerations_dict,
+                bins,
+                yvar,
+                xvars,
+                min_value,
+                max_value,
+            ],
+            share_to_global=[True, True],
+        )
+        # merge_local_histogram_numerical(local_transfers)
+
+        numerical_histogram = get_transfer_data(
+            global_run(
+                func=merge_grouped_histogram_numerical_secure2,
+                positional_args=[local_result5, bins_transfer, xvars],
+                share_to_locals=[False],
+            )
+        )
+        categorical_histogram = {}
+        print(numerical_histogram["numerical_bins"])
+
+    return_list = []
+    if yvar in nominal_vars:
+        categorical_histogram1 = Histogram(
+            var=yvar,
+            bins=list(sorted(enumerations_dict[yvar].keys())),
+            counts=categorical_histogram["categorical_histogram"],
+        )
+        return_list.append(categorical_histogram1)
+        if xvars:
+            for i, x_variable in enumerate(xvars):
+                possible_groups = sorted(enumerations_dict[x_variable].keys())
```

Comment:
Do not sort enumerations as some are already sorted, but not in alphabetical order. E.g. `agegroup` has enums `[-50y, 50-59y, 60-69y, 70-79y, +80y]`. You should use the order already present in the metadata.